### DISCUSSION TITLE
Is it possible to embed a PDF as a section on a page

### DISCUSSION THREAD

**Discussion by BeMoreDog** [pdfcpu/pdfcpu#983] [Q&A]
Hi.  Thanks for such a wonderful tool...

So I know I can merge PDFs but I'm wondering if I can take an existing page of PDF `1` (which might be smaller than the destination PDF) and add it to a `Destination` PDF.  
I suppose you could think of this as similar to embedding an image but instead of it being an image it would be a PDF that is embedded.

I suspect this isn't currently possible but is there anything technically stopping this from being achievable?

**Comment by hhrutter** [comment]
Hello!

You can attach files to a PDF and use it as a container:

pdfcpu has two commands at disposal for you:

https://pdfcpu.io/portfolio/portfolio

https://pdfcpu.io/attach/attach

**Reply by BeMoreDog** [reply]
Thanks for replying.  They don't quite achieve what I hope to do which is not to attach another PDF but instead to use another PDF as a snippet to insert visually into a page.  For example, one use could be if I have a preview of a web site that headless chrome generates as a PDF, I'd like to display that as if it were an image but I'd like to retain it's selectable text by keeping it in a PDF format, not show it as an image.

**Reply by petervwyatt** [reply]
Since PDF 1.4, there has been a special kind of XObject called a Reference XObject - see 8.10.4 in ISO 32000-2 - that "_enable[s] one PDF document to import content from another_".

However this is very advanced functionality and not implemented by all viewers, as it also has security and content trust implications as the content can be completely external to the containing PDF. It is typically used only in closed workflows.

If your use-case is for print (such as late binding of data), then there is also PDF/VCR (ISO 16613) - see https://pdfa.org/resource/iso-16613-pdfvcr/ which has similar capabilities.

**Reply by BeMoreDog** [reply]
Thanks @petervwyatt that's very helpful - Reference XObject looks like it could achieve what I'm interested in and my use case would be a closed workflow.  I see too that there is a fallback to a proxy which in my case could be a low res image - if I'm understanding correctly readers that don't support the Reference XObject would show this fallback, is that correct?  Do you know if there is there anywhere to find out which of the most common PDF readers support this?

**Reply by petervwyatt** [reply]
Yes - the proxy is designed to be the fallback and will work in all viewers that support Form XObjects - and that's effectively everything.

And, sorry, I don't know how to find out which viewers do and do not support with any limitations (likely there will be permission settings for trusted locations, etc.). Be best to contact vendors or OSS communities and/or create a simple test file and manually test within your closed environment. 

Another similar option which used to be used in print-centric environments is 14.11.7 Open prepress interface (OPI). This is now out-of-favour but I think some print workflow software may still support it, which might be another option for you.

**Comment by hhrutter** [comment]
Thanks for the clarification and thanks @petervwyatt for the standards background.

pdfcpu currently does not support creating Reference XObjects for visually embedding content from one PDF into another PDF page while preserving the referenced PDF content as PDF content.

For most workflows pdfcpu supports merging, n-up/booklet-style page composition, stamping/watermarking, and embedded file attachments/portfolios, but those are different mechanisms. Reference XObjects are a more specialized PDF feature with viewer-support and trust/security implications, so this is currently outside pdfcpu’s supported feature set.

For a closed workflow, the practical recommendation is to create a small Reference XObject test file and validate it against the exact viewers/processors used in that workflow. For broad distribution, relying on this feature is likely risky unless viewer support is controlled.

I am going to mark this as answered for now.
